odoo/tools/safe_eval.py

```diff
@@ -349,20 +798,31 @@ def safe_eval(expr, globals_dict=None, locals_dict=None, mode="eval", nocopy=Fal
         if locals_dict is not None:
             locals_dict = dict(locals_dict)
 
-    check_values(globals_dict)
-    check_values(locals_dict)
+    code_checker = CodeChecker(sandboxed_instances, sandboxed_types, ast_subset)
+    safe_env = code_checker.get_environment()
 
     if globals_dict is None:
         globals_dict = {}
 
-    globals_dict['__builtins__'] = _BUILTINS
-    if locals_builtins:
-        if locals_dict is None:
-            locals_dict = {}
-        locals_dict.update(_BUILTINS)
-    c = test_expr(expr, _SAFE_OPCODES, mode=mode, filename=filename)
+    if locals_dict is None:
+        locals_dict = {}
+
+
+    if isinstance(expr, bytes):
+        expr = expr.decode()
+
+    code = code_checker.visit(ast.parse(expr.strip(), filename=filename or ""))
+    c = ast.unparse(code)
+
+    eval_globals_dict = globals_dict.copy()
+    eval_globals_dict.update(safe_env)
+    eval_globals_dict['__builtins__'] = _BUILTINS.copy()
+
+    eval_locals_dict = locals_dict.copy()
```

Comment:
be carreful we don't put weird stuff in it

Reply:
btw is there any test about it ? :/

because tests are green, but I can do that

```py
safe_eval('[i for i in (1, 2)][0]', {}, {'__type_checker': lambda *args, **kwargs: print('bouh')})
```

Reply:
(globals are safe because we overwrite the dict, but not the locals)